# DATA PREPROCESSING
This notebook will capture the cleaning, feature engineering and feature selection part of the week 3 deliverable. The original dataset given in week 1 will be used.

## Objectives
The aim is to prepare the dataset for predictive modelling and churn analysis.


In [2]:
#reading the raw file
import pandas as pd
raw_path=r"C:\Users\maruf\SLU Opportunity Wise Data-1710158595043.xlsx"
raw = pd.read_excel(raw_path)
#Display the first 10 rows of the dataframe
raw.head(10)


Learner SignUp DateTime                  Opportunity Id  \
0     06/14/2023 12:30:35  00000000-0GN2-A0AY-7XK8-C5FZPP   
1     2023-01-05 05:29:16  00000000-0GN2-A0AY-7XK8-C5FZPP   
2     2023-09-04 20:35:08  00000000-0GN2-A0AY-7XK8-C5FZPP   
3     08/29/2023 05:20:03  00000000-0GN2-A0AY-7XK8-C5FZPP   
4     2023-06-01 15:26:36  00000000-0GN2-A0AY-7XK8-C5FZPP   
5     2024-02-03 19:16:07  00000000-0GN2-A0AY-7XK8-C5FZPP   
6     05/31/2023 07:11:04  00000000-0GN2-A0AY-7XK8-C5FZPP   
7     07/22/2023 07:30:25  00000000-0GN2-A0AY-7XK8-C5FZPP   
8     03/20/2023 16:36:30  00000000-0GN2-A0AY-7XK8-C5FZPP   
9     2023-11-05 03:40:33  00000000-0GN2-A0AY-7XK8-C5FZPP   

                                    Opportunity Name Opportunity Category  \
0  Career Essentials: Getting Started with Your P...               Course   
1  Career Essentials: Getting Started with Your P...               Course   
2  Career Essentials: Getting Started with Your P...               Course   
3  Career Essentials: Getting Started with Your P...               Course   
4  Career Essentials: Getting Started with Your P...               Course   
5  Career Essentials: Getting Started with Your P...               Course   
6  Career Essentials: Getting Started with Your P...               Course   
7  Career Essentials: Getting Started with Your P...               Course   
8  Career Essentials: Getting Started with Your P...               Course   
9  Career Essentials: Getting Started with Your P...               Course   

  Opportunity End Date        First Name        Date of Birth  Gender  \
0  06/29/2024 18:52:39             Faria  2001-12-01 00:00:00  Female   
1  06/29/2024 18:52:39          Poojitha           08/16/2000  Female   
2  06/29/2024 18:52:39          Emmanuel           01/27/2002    Male   
3  06/29/2024 18:52:39  Amrutha Varshini  1999-01-11 00:00:00  Female   
4  06/29/2024 18:52:39    Vinay Varshith           04/19/2000    Male   
5  06/29/2024 18:52:39               Mor  1996-12-05 00:00:00    Male   
6  06/29/2024 18:52:39           Fardeen  2001-09-09 00:00:00    Male   
7  06/29/2024 18:52:39             Gauri           02/27/2006  Female   
8  06/29/2024 18:52:39         SIDDHARTH           12/22/2005    Male   
9  06/29/2024 18:52:39          Vanshika           06/26/1998  Female   

                Country                     Institution Name  \
0              Pakistan                                Nwihs   
1                 India                          SAINT LOUIS   
2         United States     Illinois Institute of Technology   
3         United States               Saint Louis University   
4         United States               Saint Louis University   
5                 India               Saint Louis University   
6                 India     Illinois Institute of Technology   
7  United Arab Emirates  GEMS New Millennium School Al Khail   
8                 India                       Ashoka Academy   
9                 India     Illinois Institute of Technology   

                Current/Intended Major    Entry created at Status Description  \
0                            Radiology 2024-11-03 12:01:41            Started   
1                  Information Systems 2024-11-03 12:01:41            Started   
2                     Computer Science 2024-11-03 12:01:41            Started   
3                  Information Systems 2024-11-03 12:01:41     Team Allocated   
4                     Computer Science 2024-11-03 12:01:41            Started   
5               Mechanical Engineering 2024-11-03 12:01:41         Waitlisted   
6     Computer Science and Engineering 2024-11-03 12:01:41           Withdraw   
7              Artificial Intelligence 2024-11-03 12:01:41     Team Allocated   
8  Robotics and Automation Engineering 2024-11-03 12:01:41            Started   
9                     Computer Science 2024-11-03 12:01:41     Team Allocated   

   Status Code           Apply Date Opportunity Start Date  
0         1080  06/1

In [3]:
#get the size of dataset
raw.shape

(8558, 16)

- check the datatypes of each column

In [38]:
#Display the datatypes of each column in the dataframe
raw.dtypes

Learner SignUp DateTime            object
Opportunity Id                     object
Opportunity Name                   object
Opportunity Category               object
Opportunity End Date               object
First Name                         object
Date of Birth                      object
Gender                             object
Country                            object
Institution Name                   object
Current/Intended Major             object
Entry created at           datetime64[ns]
Status Description                 object
Status Code                         int64
Apply Date                         object
Opportunity Start Date             object
dtype: object

**Data Cleaning**

In [4]:
# Check data quality issues
print("Data Quality Issues:")
print(f"Missing values:\n{raw.isnull().sum()}")
print(f"\nDuplicate rows: {raw.duplicated().sum()}")

# Check for inconsistent categorical values
print(f"\nUnique values in categorical columns:")
categorical_cols = ['Opportunity Name','Gender', 'Country','Opportunity Category','Institution Name', 'Current/Intended Major', 'Status Description','Status Code']
for col in categorical_cols:
    if col in cp.columns:
        print(f"{col}: {cp[col].unique()}")


Data Quality Issues:
Missing values:
Learner SignUp DateTime       0
Opportunity Id                0
Opportunity Name              0
Opportunity Category          0
Opportunity End Date          0
First Name                    0
Date of Birth                 0
Gender                        0
Country                       0
Institution Name              5
Current/Intended Major        5
Entry created at              0
Status Description            0
Status Code                   0
Apply Date                    0
Opportunity Start Date     3794
dtype: int64

Duplicate rows: 0

Unique values in categorical columns:


NameError: name 'cp' is not defined

In [13]:
raw['Status Description'].value_counts()

Status Description
Rejected          3569
Team Allocated    3276
Started            767
Dropped Out        617
Waitlisted         109
Applied            105
Withdraw            86
Rewards Award       29
Name: count, dtype: int64

***Handling missing values***
- For Institution name and Current/Intended Major rows with missing values are dropped
- For Opportunity Start Date having 3794 empty rows are replaced with a place holder

In [5]:
#inputting unknown is some missing values (handling missing data)
raw['Opportunity Start Date']=raw['Opportunity Start Date'].fillna('unknown')
#Drop rows with null values from the dataframe (handling missing data)
data=raw.dropna()
#get the size to ensure accuracy
data.shape

(8548, 16)

***Handling columns formats***
- The dataset contain categorical columns and date columns

In [6]:
#converting all columns to strings
data=data.astype(str)
print(data.dtypes)

Learner SignUp DateTime    object
Opportunity Id             object
Opportunity Name           object
Opportunity Category       object
Opportunity End Date       object
First Name                 object
Date of Birth              object
Gender                     object
Country                    object
Institution Name           object
Current/Intended Major     object
Entry created at           object
Status Description         object
Status Code                object
Apply Date                 object
Opportunity Start Date     object
dtype: object


In [7]:
# Convert only the true date columns to datetime
date_columns = [
    "Learner SignUp DateTime",
    "Opportunity End Date",
    "Date of Birth",
    "Entry created at",
    "Apply Date",
    "Opportunity Start Date"
]

for col in date_columns:
    data[col] = pd.to_datetime(data[col], format='mixed', errors='coerce')

#cp1["Date of Birth"] = pd.to_datetime(cp1["Date of Birth"], format='mixed', errors='coerce').dt.date


print(data.dtypes)
print(data.shape)



Learner SignUp DateTime    datetime64[ns]
Opportunity Id                     object
Opportunity Name                   object
Opportunity Category               object
Opportunity End Date       datetime64[ns]
First Name                         object
Date of Birth              datetime64[ns]
Gender                             object
Country                            object
Institution Name                   object
Current/Intended Major             object
Entry created at           datetime64[ns]
Status Description                 object
Status Code                        object
Apply Date                 datetime64[ns]
Opportunity Start Date     datetime64[ns]
dtype: object
(8548, 16)


***Handling values errors ***
- some values are written inconsistently

In [67]:
# getting the columns with issues
#print(data['Country'].unique())
#print(data['Institution Name'].unique())

In [8]:
#Handling Inconsistent country names based on unique vallues details:
data['Country'] = data['Country'].str.replace('U.S.', 'United States')
data['Country'] = data['Country'].str.replace('Iran, Islamic Republic of Persian Gulf', 'Iran')
data['Country'] = data['Country'].str.replace('Iran  Islamic Republic of Persian Gulf', 'Iran')

#Handling Inconsistent Institution names based on unique vallues details:
data['Institution Name'] = data['Institution Name'].str.replace('saint louis university', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].str.replace('Saint Louis university', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].str.replace('Saint louis university', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].str.replace('SAINT LOUIS UNIVERSITY', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].str.replace('Saint louis University', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].str.replace('St Louis University', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].str.replace('Saint Louis', 'Saint Louis University')
# noticed the name coming as Saint Louis University University so tried to adjust
data['Institution Name'] = data['Institution Name'].str.replace('Saint Louis University University', 'Saint Louis')

data['Institution Name'] = data['Institution Name'].str.replace('saint Louis University', 'Saint Louis')
data['Institution Name'] = data['Institution Name'].replace('Saint Louis', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].replace('saint Louis university', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].replace('saint louis', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].replace('saint louis University', 'Saint Louis University')
data['Institution Name'] = data['Institution Name'].replace('SAINT LOUIS', 'Saint Louis University')
print(data['Institution Name'].value_counts())
  


Institution Name
Saint Louis University                                4321
Illinois Institute of Technology                       103
Webster University                                      46
University of Ibadan                                    26
Kwame Nkrumah University of Science and Technology      22
                                                      ... 
Federal University of Petroleum Resources  Effurun       1
ADDIS ABABA INSTITUTE OF TECHNOLOGY                      1
Marwadi University                                       1
RMKCET                                                   1
NIIT  Accra                                              1
Name: count, Length: 2073, dtype: int64


In [9]:
# converting DOB to disregard time
data["Date of Birth"] = pd.to_datetime(data["Date of Birth"], format='mixed', errors='coerce').dt.date
#inputting place holder to avoid dropping rows
data['Opportunity Start Date']=data['Opportunity Start Date'].fillna('NA')


In [10]:
# Drop null rows 
data_cleaned = data
data_cleaned = data_cleaned.dropna()
data_cleaned.shape


(6839, 16)

In [11]:
# Save cleaned file

output_path0 = r"C:\Users\maruf\T1_cleaned_data_week3.xlsx"

data_cleaned.to_excel(output_path0, index=False)

print(f"✅ Cleaned file saved successfully at: {output_path}")
print(data_cleaned.shape)

NameError: name 'output_path' is not defined

**Feature Engineering**


In [13]:
#reading and viewing the cleaned file
cleaned_data=pd.read_excel(output_path0)
cleaned_data.head()

Learner SignUp DateTime                  Opportunity Id  \
0     2023-06-14 12:30:35  00000000-0GN2-A0AY-7XK8-C5FZPP   
1     2023-01-05 05:29:16  00000000-0GN2-A0AY-7XK8-C5FZPP   
2     2023-08-29 05:20:03  00000000-0GN2-A0AY-7XK8-C5FZPP   
3     2023-06-01 15:26:36  00000000-0GN2-A0AY-7XK8-C5FZPP   
4     2024-02-03 19:16:07  00000000-0GN2-A0AY-7XK8-C5FZPP   

                                    Opportunity Name Opportunity Category  \
0  Career Essentials: Getting Started with Your P...               Course   
1  Career Essentials: Getting Started with Your P...               Course   
2  Career Essentials: Getting Started with Your P...               Course   
3  Career Essentials: Getting Started with Your P...               Course   
4  Career Essentials: Getting Started with Your P...               Course   

  Opportunity End Date        First Name Date of Birth  Gender        Country  \
0  2024-06-29 18:52:39             Faria    2001-12-01  Female       Pakistan   
1  2024-06-29 18:52:39          Poojitha    2000-08-16  Female          India   
2  2024-06-29 18:52:39  Amrutha Varshini    1999-01-11  Female  United States   
3  2024-06-29 18:52:39    Vinay Varshith    2000-04-19    Male  United States   
4  2024-06-29 18:52:39               Mor    1996-12-05    Male          India   

         Institution Name  Current/Intended Major    Entry created at  \
0                   Nwihs               Radiology 2024-11-03 12:01:41   
1  Saint Louis University     Information Systems 2024-11-03 12:01:41   
2  Saint Louis University     Information Systems 2024-11-03 12:01:41   
3  Saint Louis University        Computer Science 2024-11-03 12:01:41   
4  Saint Louis University  Mechanical Engineering 2024-11-03 12:01:41   

  Status Description  Status Code          Apply Date Opportunity Start Date  
0            Started         1080 2023-06-14 12:36:09    2022-03-11 18:30:39  
1            Started         1080 2023-01-05 06:08:21    2022-03-11 18:30:39  
2     Team Allocated         1070 2023-09-10 22:02:42    2022-03-11 18:30:39  
3            Started         1080 2023-06-01 15:40:10    2022-03-11 18:30:39  
4         Waitlisted         1040 2024-02-03 20:30:35                    NaT

In [ ]:
#cleaned_data['Current/Intended Major'].unique()

In [14]:
#creating Major Category (STEM/non-STEM/Others) using unique values
stem=['Radiology', 'Information Systems', 'Computer Science', 'Epidemiology', 'Petroleum Engineering', 
       'Mechanical Engineering', 'Computer Science and Engineering','Pre Engg', 'Civil and Structural Engineering',
       'Artificial Intelligence', 'Robotics and Automation Engineering','Program in Exploratory Studies', 'Architectural Engineering',
       'Data Visualization', 'Public Health','Biotechnology', 'Airline Pilot', 'Food Safety and Technology','Medical Laboratory Science',
       'Systems Engineering','Geology', 'Neuroscience', 'Data Analysis', 'Bsc Nursing', 'Information Engineering','Electronics and Computer Engineering',
       'Architecture', 'Computer Science and Information Systems','Aeronautics', 'Engineers Physics', 'Construction Engineering and Management',
       'Soil Science and Technology', 'Business','Bs Forestry and Wildlife Management', 'Medical Laboratory','Occupational Therapy', 
       'Analytics', 'Biology', 'Economics', 'Mathematics', 'Bioinformatics','Data Science and Artificial Intelligence','Aeronautical Engineering',
       'Msc cyber forensics and security', 'Statistics and Operational Research','Master in VLSI and Microelectronics',
        'Biomedical Engineering', 'Data Science', 'Statistics','Economics and Econometrics', 'Computer and Infromation Sciences',
       'Bsc in Computer Science', 'Electronics and Communication', 'Computer Information Systems','Post Graduate Diploma in Public Health',
       'Electrical and Electronic Engineering', 'Electrical and Computer Engineering','Management Information Systems', 'Project Management',
       'Medicine', 'Information', 'Information Technology', 'Monitoring and Evaluation', 'Electrical Engineer','Social Development and Policy',
       'Actuarial Mathematics', 'Software Engineering', 'Linguistics', 'Voice Performance', 'Bachelor of Laws', 'Automotive Engineering',
       'Library and Information Management', 'Biological Sciences', 'Urban and Housing Development', 'Cyber Security', 'Data Analytics',
       'Computer Engineering', 'Environmental Sciences','Industrial Engineering', 'Business Analytics','Functional Engineering',
       'Artificial Intelligence and Machine learning','Forensic', 'Forensic Science', 'SUPPLY CHAIN MANAGEMENT',
       'Agriculture and Forestry', 'Pure and Applied Physics','Rehabilitation and Mental Health Counseling','MS in Information Science Machine Learning',
      'Applied Geophysics','Electronics and Computers Engineering', 'Computer and Information Sciences', 'B TECH', 'Accountancy',
       'Information technology Management', 'Archaeology', 'Logistics', 'ENGINEERING Management', 'MASTERS in INFORMATION SYSTEMS',
       'Bnsc Nursing Science', 'Computer Science Engineering', 'Aviation Management', 'Land Economy', 'ACCOUNTING',
       'MSc in Computational Decision Science and Operations Research', 'Biochemistry', 'Information Technology and Management',
       'Civil Engineering', 'Electrical Engineering', 'Cybersecurity', 'Computer Science with Internet of Things',
       'Pharmacy and Pharmacology', 'Nutrition Science', 'Metallurgy and Materials Engineering', 'Finance', 'Geography',
       'Information technology and management', 'Dentistry', 'Diploma in Visual Effects', 'Digital Media', 'Computer and Information Sciencess',
       'Environmental Sustainability', 'Electronics and Computer Science','MSc in Information technology','Tech and Business Management',
       'Communication and Software Engineer', 'Respiratory Therapy','Health Information Management', 'Veterinary Science',
       'Health Education','Food Science', 'Psychology', 'MS in Supply Chain Management', 'Financial Technology', 'Information technology', 
'Health Data Science', 'Electronics and Communication Engineering', 'Quantity Surveying','Masters in Information Systems', 
       'Masters in Computer Science','Environmental and Safety Engineering','ESports and Game Development', 'Could Computing',
       'Network and Information Security', 'MBA','Computer Applications', 'Mechanical and Aerospace Engineering',
       'Mechatronics Engineering', 'Supply Chain Management', 'Nursing','Animal Production and Health','Artificial Intelligence and Robotics',
       'Chemistry','Btech','Chemical Engineering',  'Physics and Astronomy','Earth and Marine Sciences', 'ANALYTICS',
       'Radiation Therapy',  'Engineering','Aerospace Engineering', 'Bachelor of Engineering', 'Information Science', 
       'Associate of Science', 'Marketing Analytics', 'Medical Sciences','Finance and Banking', 'Agricultural Extension and Education',
       'Applied Financial Economics', 'DOCTOR of PHARMACY', 'Bsc Agri', 'Telecommunication Engineering','Computing', 
       'Anatomy and Physiology','Artificial Intelligence and Data Science','Electronics and Computer Science Engineering',
       'Physics','MS Biostatistics and Health Analytics','Computer Science and Business Systems',  'Computer', 'Manufacturing Engineering',
       'Diploma in Automation and Robotics', 'Archeology and Cinematography', 'Carpentry and Joinery','Data Analyst','B Tech', 
      'Computer science engineering', 'Ms in Analytics','Food Process Engineering','Data Sceince', 'Animal Production',
       'COMPUTER SCIENCE and DESIGN', 'Business Informatics','Cyber Forensics and Security', 'Management Science and Analytics',
       'Actuarial Science', 'Mineral and Mining Engineering', 'People Analytics', 'Autonomous Systems and Robotics', 'Nurse',
       'Business Education', 'Data science', 'Remote Sensing','IT',  'Electronics and Telecommunication Engineering', 'MECHANICAL',
       'Textile Engineering', 'Machine Learning', 'Analysis','Healthcare Management','Physical Therapy', 'Human Nutrition and Dietetics',
       'Health Sciences', 'Microbiology', 'Information Studies', 'Physiotherapy', 'Statistics and Computeri Science', 'Radiography', 
       'MS in Cybersecurity', 'Msc in Disaster Management and Climate Sustainability Studies','Cyber Security and Digital Forensics', 
       'Procurement and Supply Chain Management', 'Information systems','Fisheries',  'Computer Science and Design', 'BSC Cs',
       'Msters in Healthcare Administration','Computer Networking and IT Security','Health Informatics','Bachelor Degree in Nursing',
       'Geoinformatics and Geospatial Analytics', 'Informatics','Electronics and Instrumentation',
       'Bachelor of Computer', 'Accounting', 'ECONOMICS and FINANCE','Geographic Information Science',
       'Agricultural Economics and Farm Management', 'General Science', 'Working As Software Devloper',]

nstem=['Business Administration','Business and Management Studies','Accounting and Finance','Human Resource', 'Law','Journalism',
       'Secretarial','Human Resources',  'Philosophy','Entrepreneurship','Communication and Media Studies', 'BCOM International Business', 
       'Law and Legal Studies', 'Theology or Divinity and Religious Studies', 'International Business Management','Advertising', 'History',
       'Politics', 'Marketing', 'English', 'English Language and Literature','Communication','English Linguistics','Social Policy and Administration',
       'Hospitality and Leisure Management', 'Mass Communication', 'Commerce','Digital Marketing','Political Science', 'Not Know Ab',
       'International Relations',  'Public Administration', 'Education and Training','Sociology', 'Business Chinese', 'Music', 'Education Management',
          'Management', 'Procurement', 'Art and Design','Development Studies', 'General Arts','Medical Marketing or Medical Administration',
     'Guidance and Counseling', 'Arts', 'Home Economics', 'Performing Arts','Textile', 'Education and Geography','General Management',
        'Development Economics', 'Film Making','International Development',  'International Business', 'Communication and Language Arts Education',
       'Early Childhood Education','Islamic Studies', 'Multimedia Communication','Entrepreneurship Studies','Advertising and Packaging',
       'American Sign Language Interpretation', 'Communications', 'Brand Strategy', 'Leadership and Organizational development',
       'Education Administration',]

others=['Oth','I Am Major','Other', 'Student', 'Eee', 'Social Work','Pol','Fresher','Faizan','Degree','BSIT','Job', 'Leadership','Gamer', 'Ot',
        'BBA','MCA',  'No', 'Aa','Community Health Officer', 'Social S','Ba Psych', 'Shaista', 'Ha', 'Unemployed','Dropped Out',
       'After Bachelor', 'MMG', 'Goodcreativity','Bachelor','Diploma','BCA', 'TE','Manager','Yoganand Sir', 'High School Certificate', 'BAF',
       'Lower Credit','Certification','Pursuing MCA','Othe', 'Honours', 'ISE','Otheraassss','High School', 'To Study', 'General Practitioner',
       'Research','Non', 'Sdada', 'Ghj', 'Na', 'General', 'Bcom','Masters', 'B Sc','Www','Already Graduation Completed','GENERAL MANAGEMENT', 
       'High School Student', 'Pv','Home Tutor', 'Cycw', 'Nil','Pos Service']
#creating empty column
cleaned_data['major_category']=''
# assigning conditions
cleaned_data.loc[cleaned_data['Current/Intended Major'].isin(stem), 'major_category']='STEM'
cleaned_data.loc[cleaned_data['Current/Intended Major'].isin(nstem), 'major_category']='non-STEM'
cleaned_data.loc[cleaned_data['Current/Intended Major'].isin(others), 'major_category']='Others'


In [303]:
cleaned_data.Country.unique()

array(['Pakistan', 'India', 'United States', 'United Arab Emirates',
       'Nigeria', 'Egypt', 'Nepal', 'Kenya', 'Ghana', 'Zambia', 'Morocco',
       'Ethiopia', 'Zimbabwe', 'Uganda', 'Indonesia', 'Cameroon', 'Yemen',
       'China', 'Bangladesh', 'Congo', 'Liberia', 'United Kingdom',
       'Vietnam', 'Japan', 'Rwanda', 'Gambia', 'Philippines', 'Australia',
       'Somalia', 'Sierra Leone', 'Lebanon', 'Botswana', 'Iraq',
       'Uzbekistan', 'Turkey', 'Honduras',
       'Tanzania, United Republic of Tanzania',
       'British Indian Ocean Territory', 'France', 'Belarus', 'Algeria',
       'Korea, Republic of South Korea', 'Mauritius', 'Tunisia',
       'Kazakhstan', 'Peru', 'Brazil', 'Ukraine', 'South Africa',
       'Germany', 'Namibia', 'Iran', 'American Samoa',
       'Falkland Islands (Malvinas)', 'Saudi Arabia', 'Azerbaijan',
       'Dominican Republic', 'Lesotho', 'Malaysia', 'Singapore',
       'Libyan Arab Jamahiriya', "Cote d'Ivoire", 'Afghanistan', 'Canada',
       'Bhutan'

In [15]:
#creating country regions countries unique values 
Africa=['Algeria', 'Botswana', 'Cameroon', 'Congo', "Cote d'Ivoire", 'Egypt', 'Ethiopia', 'Gambia', 'Ghana', 'Kenya', 'Lesotho', 
'Liberia', 'Libyan Arab Jamahiriya', 'Mauritius', 'Morocco', 'Namibia', 'Nigeria', 'Rwanda', 'Sierra Leone', 'Somalia',
 'South Africa', 'Tanzania, United Republic of Tanzania', 'Tunisia', 'Uganda', 'Zambia', 'Zimbabwe']
Asia=['Afghanistan', 'Azerbaijan', 'Bangladesh', 'Bhutan', 'British Indian Ocean Territory', 'China', 'India', 'Indonesia', 'Iran',
 'Iraq', 'Japan', 'Kazakhstan', 'Korea, Republic of South Korea', 'Lebanon', 'Malaysia', 'Nepal', 'Pakistan', 'Philippines',
 'Saudi Arabia', 'Singapore', 'Turkey', 'United Arab Emirates', 'Uzbekistan', 'Vietnam', 'Yemen']
Europe=['Belarus', 'France', 'Germany', 'Spain', 'Ukraine', 'United Kingdom']
North_America=['Canada', 'Dominican Republic', 'Honduras', 'United States']
Oceania=['American Samoa', 'Australia']
South_America=['Brazil', 'Falkland Islands (Malvinas)', 'Peru']
#creating empty column
cleaned_data['region']=''
# assigning conditions
cleaned_data.loc[cleaned_data['Country'].isin(Africa), 'region']='Africa'
cleaned_data.loc[cleaned_data['Country'].isin(Asia), 'region']='Asia'
cleaned_data.loc[cleaned_data['Country'].isin(Europe), 'region']='Europe'
cleaned_data.loc[cleaned_data['Country'].isin(North_America), 'region']='North America'
cleaned_data.loc[cleaned_data['Country'].isin(South_America), 'region']='South_America'
cleaned_data.loc[cleaned_data['Country'].isin(Oceania), 'region']='Oceania'


In [16]:
# replacing all but one university due to its high count
university_to_keep = 'Saint Louis University'

# Assign the new value 'Others' to other rows
cleaned_data.loc[cleaned_data['Institution Name'] != university_to_keep, 'Institution Name'] = 'Others'

print("\nDataFrame after replacement:")
cleaned_data.head()


DataFrame after replacement:


Learner SignUp DateTime                  Opportunity Id  \
0     2023-06-14 12:30:35  00000000-0GN2-A0AY-7XK8-C5FZPP   
1     2023-01-05 05:29:16  00000000-0GN2-A0AY-7XK8-C5FZPP   
2     2023-08-29 05:20:03  00000000-0GN2-A0AY-7XK8-C5FZPP   
3     2023-06-01 15:26:36  00000000-0GN2-A0AY-7XK8-C5FZPP   
4     2024-02-03 19:16:07  00000000-0GN2-A0AY-7XK8-C5FZPP   

                                    Opportunity Name Opportunity Category  \
0  Career Essentials: Getting Started with Your P...               Course   
1  Career Essentials: Getting Started with Your P...               Course   
2  Career Essentials: Getting Started with Your P...               Course   
3  Career Essentials: Getting Started with Your P...               Course   
4  Career Essentials: Getting Started with Your P...               Course   

  Opportunity End Date        First Name Date of Birth  Gender        Country  \
0  2024-06-29 18:52:39             Faria    2001-12-01  Female       Pakistan   
1  2024-06-29 18:52:39          Poojitha    2000-08-16  Female          India   
2  2024-06-29 18:52:39  Amrutha Varshini    1999-01-11  Female  United States   
3  2024-06-29 18:52:39    Vinay Varshith    2000-04-19    Male  United States   
4  2024-06-29 18:52:39               Mor    1996-12-05    Male          India   

         Institution Name  Current/Intended Major    Entry created at  \
0                  Others               Radiology 2024-11-03 12:01:41   
1  Saint Louis University     Information Systems 2024-11-03 12:01:41   
2  Saint Louis University     Information Systems 2024-11-03 12:01:41   
3  Saint Louis University        Computer Science 2024-11-03 12:01:41   
4  Saint Louis University  Mechanical Engineering 2024-11-03 12:01:41   

  Status Description  Status Code          Apply Date Opportunity Start Date  \
0            Started         1080 2023-06-14 12:36:09    2022-03-11 18:30:39   
1            Started         1080 2023-01-05 06:08:21    2022-03-11 18:30:39   
2     Team Allocated         1070 2023-09-10 22:02:42    2022-03-11 18:30:39   
3            Started         1080 2023-06-01 15:40:10    2022-03-11 18:30:39   
4         Waitlisted         1040 2024-02-03 20:30:35                    NaT   

  major_category         region  
0           STEM           Asia  
1           STEM           Asia  
2           STEM  North America  
3           STEM  North America  
4           STEM           Asia

In [305]:
cleaned_data['Status Description'].unique()

array(['Started', 'Team Allocated', 'Waitlisted', 'Withdraw',
       'Rewards Award', 'Dropped Out', 'Rejected', 'Applied'],
      dtype=object)

In [17]:
#creating dropped out column
dropped=['Dropped Out']
not_dropped=['Rewards Award','Withdraw']
#creating empty column
cleaned_data=cleaned_data.assign(dropped_out=None)
# assigning conditions
cleaned_data.loc[cleaned_data['Status Description'].isin(dropped), 'dropped_out']='Yes'
cleaned_data.loc[cleaned_data['Status Description'].isin(not_dropped), 'dropped_out']='No'
cleaned_data.head()

Learner SignUp DateTime                  Opportunity Id  \
0     2023-06-14 12:30:35  00000000-0GN2-A0AY-7XK8-C5FZPP   
1     2023-01-05 05:29:16  00000000-0GN2-A0AY-7XK8-C5FZPP   
2     2023-08-29 05:20:03  00000000-0GN2-A0AY-7XK8-C5FZPP   
3     2023-06-01 15:26:36  00000000-0GN2-A0AY-7XK8-C5FZPP   
4     2024-02-03 19:16:07  00000000-0GN2-A0AY-7XK8-C5FZPP   

                                    Opportunity Name Opportunity Category  \
0  Career Essentials: Getting Started with Your P...               Course   
1  Career Essentials: Getting Started with Your P...               Course   
2  Career Essentials: Getting Started with Your P...               Course   
3  Career Essentials: Getting Started with Your P...               Course   
4  Career Essentials: Getting Started with Your P...               Course   

  Opportunity End Date        First Name Date of Birth  Gender        Country  \
0  2024-06-29 18:52:39             Faria    2001-12-01  Female       Pakistan   
1  2024-06-29 18:52:39          Poojitha    2000-08-16  Female          India   
2  2024-06-29 18:52:39  Amrutha Varshini    1999-01-11  Female  United States   
3  2024-06-29 18:52:39    Vinay Varshith    2000-04-19    Male  United States   
4  2024-06-29 18:52:39               Mor    1996-12-05    Male          India   

         Institution Name  Current/Intended Major    Entry created at  \
0                  Others               Radiology 2024-11-03 12:01:41   
1  Saint Louis University     Information Systems 2024-11-03 12:01:41   
2  Saint Louis University     Information Systems 2024-11-03 12:01:41   
3  Saint Louis University        Computer Science 2024-11-03 12:01:41   
4  Saint Louis University  Mechanical Engineering 2024-11-03 12:01:41   

  Status Description  Status Code          Apply Date Opportunity Start Date  \
0            Started         1080 2023-06-14 12:36:09    2022-03-11 18:30:39   
1            Started         1080 2023-01-05 06:08:21    2022-03-11 18:30:39   
2     Team Allocated         1070 2023-09-10 22:02:42    2022-03-11 18:30:39   
3            Started         1080 2023-06-01 15:40:10    2022-03-11 18:30:39   
4         Waitlisted         1040 2024-02-03 20:30:35                    NaT   

  major_category         region dropped_out  
0           STEM           Asia        None  
1           STEM           Asia        None  
2           STEM  North America        None  
3           STEM  North America        None  
4           STEM           Asia        None

In [18]:
# creating age column
# Age in years: difference between today and Date of Birth
today = pd.Timestamp.today()
cleaned_data['age'] = (today - cleaned_data['Date of Birth']).dt.days // 365
# if age is all NaN (DOB not present), try to keep existing 'age' column if available
if cleaned_data['age'].isna().all() and 'age' in cleaned_data.columns:
    cleaned_data['age'] = cleaned_data['age']
cleaned_data.head()

Learner SignUp DateTime                  Opportunity Id  \
0     2023-06-14 12:30:35  00000000-0GN2-A0AY-7XK8-C5FZPP   
1     2023-01-05 05:29:16  00000000-0GN2-A0AY-7XK8-C5FZPP   
2     2023-08-29 05:20:03  00000000-0GN2-A0AY-7XK8-C5FZPP   
3     2023-06-01 15:26:36  00000000-0GN2-A0AY-7XK8-C5FZPP   
4     2024-02-03 19:16:07  00000000-0GN2-A0AY-7XK8-C5FZPP   

                                    Opportunity Name Opportunity Category  \
0  Career Essentials: Getting Started with Your P...               Course   
1  Career Essentials: Getting Started with Your P...               Course   
2  Career Essentials: Getting Started with Your P...               Course   
3  Career Essentials: Getting Started with Your P...               Course   
4  Career Essentials: Getting Started with Your P...               Course   

  Opportunity End Date        First Name Date of Birth  Gender        Country  \
0  2024-06-29 18:52:39             Faria    2001-12-01  Female       Pakistan   
1  2024-06-29 18:52:39          Poojitha    2000-08-16  Female          India   
2  2024-06-29 18:52:39  Amrutha Varshini    1999-01-11  Female  United States   
3  2024-06-29 18:52:39    Vinay Varshith    2000-04-19    Male  United States   
4  2024-06-29 18:52:39               Mor    1996-12-05    Male          India   

         Institution Name  Current/Intended Major    Entry created at  \
0                  Others               Radiology 2024-11-03 12:01:41   
1  Saint Louis University     Information Systems 2024-11-03 12:01:41   
2  Saint Louis University     Information Systems 2024-11-03 12:01:41   
3  Saint Louis University        Computer Science 2024-11-03 12:01:41   
4  Saint Louis University  Mechanical Engineering 2024-11-03 12:01:41   

  Status Description  Status Code          Apply Date Opportunity Start Date  \
0            Started         1080 2023-06-14 12:36:09    2022-03-11 18:30:39   
1            Started         1080 2023-01-05 06:08:21    2022-03-11 18:30:39   
2     Team Allocated         1070 2023-09-10 22:02:42    2022-03-11 18:30:39   
3            Started         1080 2023-06-01 15:40:10    2022-03-11 18:30:39   
4         Waitlisted         1040 2024-02-03 20:30:35                    NaT   

  major_category         region dropped_out  age  
0           STEM           Asia        None   23  
1           STEM           Asia        None   25  
2           STEM  North America        None   26  
3           STEM  North America        None   25  
4           STEM           Asia        None   28

In [19]:
#creating signup year, month and weekday
# Extract month/year from Learner SignUp DateTime and compute time in opportunity
cleaned_data['Learner SignUp DateTime'] = pd.to_datetime(cleaned_data.get('Learner SignUp DateTime'), errors='coerce')
if 'Learner SignUp DateTime' in cleaned_data.columns:
    cleaned_data['signup_month'] = cleaned_data['Learner SignUp DateTime'].dt.month
    cleaned_data['signup_year'] = cleaned_data['Learner SignUp DateTime'].dt.year
    cleaned_data['signup_weekday'] = cleaned_data['Learner SignUp DateTime'].dt.weekday
    print('Added signup_weekday, signup_month and signup_year')



Added signup_weekday, signup_month and signup_year


In [20]:
#creating opportunity wait days/engagement days

# Engagement days: difference between Opportunity Start Date and Apply Date (days). Use positive days where Start >= Apply.
cleaned_data['engagement_days'] = -(cleaned_data['Opportunity Start Date'] - cleaned_data['Apply Date']).dt.days
print('Added engagement_days')
#creating opportunity duration/time in opportunity
# Time in opportunity: difference between Opportunity End Date and Opportunity Start Date (days)
cleaned_data['Opportunity End Date'] = pd.to_datetime(cleaned_data.get('Opportunity End Date'), errors='coerce')
cleaned_data['time_in_opportunity'] = (cleaned_data['Opportunity End Date'] - cleaned_data['Opportunity Start Date']).dt.days
print('Added time_in_opportunity (days)')
print('New columns sample:')
cleaned_data[['dropped_out','signup_month','signup_year','signup_weekday','region','major_category','age','engagement_days','time_in_opportunity']].head(5)


Added engagement_days
Added time_in_opportunity (days)
New columns sample:


dropped_out  signup_month  signup_year  signup_weekday         region  \
0        None             6         2023               2           Asia   
1        None             1         2023               3           Asia   
2        None             8         2023               1  North America   
3        None             6         2023               3  North America   
4        None             2         2024               5           Asia   

  major_category  age  engagement_days  time_in_opportunity  
0           STEM   23            460.0                841.0  
1           STEM   25            300.0                841.0  
2           STEM   26            549.0                841.0  
3           STEM   25            447.0                841.0  
4           STEM   28              NaN                  NaN

In [21]:
# Export the dataframe to an Excel file
output_path = r"C:\Users\maruf\Feature_Engineered_Data.xlsx"

# index=False → removes the dataframe index column
# engine='openpyxl' ensures correct Excel writing
cleaned_data.to_excel(output_path, index=False, engine='openpyxl')

print(f"✅ DataFrame exported successfully to '{output_path}'")
cleaned_data.shape

✅ DataFrame exported successfully to 'C:\Users\maruf\Feature_Engineered_Data.xlsx'


(6839, 25)

**Splitting dataset for training and testing**

- Normalize numerical features  data (age, engagement_days, time_in_opportunity)

In [23]:
from sklearn.preprocessing import StandardScaler
num_cols = [c for c in ['age','engagement_days','time_in_opportunity'] if c in cleaned_data.columns]
if num_cols:
    scaler = StandardScaler()
    cleaned_data[num_cols] = scaler.fit_transform(cleaned_data[num_cols].fillna(0))
    print('Normalized:', num_cols)
    

Normalized: ['age', 'engagement_days', 'time_in_opportunity']


In [24]:
#dropped rows with missing dropoff data(with status-started, waitlisted, team allocated, and rejected)
df=cleaned_data.dropna(subset=['dropped_out'])
df.shape

(676, 25)

In [27]:
#selecting predicting target
y=df.dropped_out
#choosing features
features=['Opportunity Category','Gender','Status Code','major_category','age','signup_month','signup_year','Institution Name',
          'engagement_days','time_in_opportunity','signup_weekday','region']
X=df[features]
#splitting data
from sklearn.model_selection import train_test_split
train_X,test_X,train_y,test_y=train_test_split(X,y, test_size=0.4, random_state=1)

df.shape

(676, 25)

- One-hot encode categorical columns for training data: Gender, Country, Opportunity Category,major_category, region, dropped_out

In [28]:
cat_cols = [c for c in ['Gender','Opportunity Category','major_category','region','Institution Name'] if c in train_X.columns]
from sklearn.preprocessing import OneHotEncoder

# Create an instance of OneHotEncoder
ohe = OneHotEncoder(handle_unknown='ignore')

# Fit the encoder on the training data only
ohe.fit(train_X[cat_cols])

# Transform training data
X_train_encoded = ohe.transform(train_X[cat_cols])

# Get the new column names created by the encoder
ohe_feature_names = ohe.get_feature_names_out(input_features=cat_cols)

# Create dataframes from the encoded sparse matrices
X_train_ohe = pd.DataFrame(X_train_encoded.toarray(), columns=ohe_feature_names, index=train_X.index)

# Drop the original categorical columns from the dataframes
train_X = train_X.drop(columns=cat_cols)

# Concatenate the original and new encoded dataframes
train_X = pd.concat([train_X, X_train_ohe], axis=1)

# Transform test data
X_test_encoded = ohe.transform(test_X[cat_cols])
# Create dataframes from the encoded sparse matrices
X_test_ohe = pd.DataFrame(X_test_encoded.toarray(), columns=ohe_feature_names, index=test_X.index)
# Drop the original categorical columns from the dataframes
test_X = test_X.drop(columns=cat_cols)
# Concatenate the original and new encoded dataframes
test_X = pd.concat([test_X, X_test_ohe], axis=1)
test_X

Status Code       age  signup_month  signup_year  engagement_days  \
1643         1050 -0.108485            12         2023        -1.248523   
1866         1050  0.121999            12         2023        -1.232032   
1929         1050  0.582966             3         2023        -0.325027   
3224         1050 -0.569452             6         2023        -1.281505   
3206         1110 -0.569452             6         2023        -0.325027   
...           ...       ...           ...          ...              ...   
3149         1050  0.352482             7         2024        -0.325027   
3897         1050 -1.260902             1         2023        -1.524747   
2404         1050 -0.799935            12         2023        -0.823880   
2059         1050 -0.108485            12         2023        -1.223786   
3061         1050  1.274416             8         2024        -0.325027   

      time_in_opportunity  signup_weekday  Gender_Don't want to specify  \
1643            -0.373702               2                           0.0   
1866            -0.373702               5                           0.0   
1929            -0.648892               6                           0.0   
3224            -0.373702               2                           0.0   
3206            -0.648892               4                           0.0   
...                   ...             ...                           ...   
3149            -0.648892               0                           0.0   
3897            -0.373702               1                           0.0   
2404            -0.373702               3                           0.0   
2059            -0.373702               1                           0.0   
3061            -0.648892               3                           0.0   

      Gender_Female  Gender_Male  ...  major_category_STEM  \
1643            0.0          1.0  ...                  0.0   
1866            0.0          1.0  ...                  1.0   
1929            0.0          1.0  ...                  0.0   
3224            0.0          1.0  ...                  1.0   
3206            0.0          1.0  ...                  1.0   
...             ...          ...  ...                  ...   
3149            1.0          0.0  ...                  1.0   
3897            0.0          1.0  ...                  1.0   
2404            1.0          0.0  ...                  1.0   
2059            1.0          0.0  ...                  1.0   
3061            0.0          1.0  ...                  1.0   

      major_category_non-STEM  region_Africa  region_Asia  region_Europe  \
1643                      1.0            0.0          1.0            0.0   
1866                      0.0            0.0          0.0            0.0   
1929                      1.0            1.0          0.0            0.0   
3224                      0.0            0.0          1.0            0.0   
3206                      0.0            0.0          1.0            0.0   
...                       ...            ...          ...            ...   
3149                      0.0            0.0          1.0            0.0   
3897                      0.0            0.0          1.0            0.0   
2404                      0.0            0.0          1.0            0.0   
2059                      0.0            0.0          0.0            0.0   
3061                      0.0            1.0          0.0            0.0   

      region_North America  region_Oceania  region_South_America  \
1643                   0.0             0.0                   0.0   
1866                   1.0             0.0                   0.0   
1929                   0.0             0.0                   0.0   
3224                   0.0             0.0                   0.0   
3206                   0.0             0.0                   0.0   
...                    ...             ...                   ...   
3149                   0.0             0.0                   0.0   
38

In [29]:

# Export the  testing target to an Excel file
output_path1a =  r"C:\Users\maruf\test_y.xlsx"
output_path1b =  r"C:\Users\maruf\train_y.xlsx"
output_path2a =  r"C:\Users\maruf\testX_Data.xlsx"
output_path2b=  r"C:\Users\maruf\trainX_Data.xlsx"

# index=False → removes the dataframe index column
# engine='openpyxl' ensures correct Excel writing
test_y.to_excel(output_path1a, index=False, engine='openpyxl')
train_y.to_excel(output_path1b, index=False, engine='openpyxl')
test_X.to_excel(output_path2a, index=False, engine='openpyxl')
train_X.to_excel(output_path2b, index=False, engine='openpyxl')

print(f"✅ Model test target and features '{output_path1a}', '{output_path2a}'")
print(f"✅ Model train target and features to '{output_path1b}', '{output_path2b}'")

✅ Model test target and features 'C:\Users\maruf\test_y.xlsx', 'C:\Users\maruf\testX_Data.xlsx'
✅ Model train target and features to 'C:\Users\maruf\train_y.xlsx', 'C:\Users\maruf\trainX_Data.xlsx'
